pallets/gear-program/src/pause.rs

```diff
@@ -0,0 +1,119 @@
+// This file is part of Gear.
+
+// Copyright (C) 2022 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use super::*;
+use codec::{Decode, Encode};
+use common::{self, QueuedDispatch};
+use frame_support::storage::PrefixIterator;
+use scale_info::TypeInfo;
+use sp_std::collections::btree_map::BTreeMap;
+
+#[derive(Clone, Debug, PartialEq, Decode, Encode, TypeInfo)]
+pub(super) struct PausedProgram {
+    program_id: H256,
+    program: common::ActiveProgram,
+    pages_hash: H256,
+    wait_list: Vec<QueuedDispatch>,
+    waiting_init: Vec<H256>,
+}
+
+fn decode_dispatch_tuple(_key: &[u8], value: &[u8]) -> Result<(QueuedDispatch, u32), codec::Error> {
+    <(QueuedDispatch, u32)>::decode(&mut &*value)
+}
+
+fn memory_pages_hash(pages: &BTreeMap<u32, Vec<u8>>) -> H256 {
+    pages.using_encoded(sp_io::hashing::blake2_256).into()
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum PauseError {
```

Comment:
Any particular reason you opted out of the Substrate idiomatic way of defining errors in pallets via `pallet::error` macro?
```
#[pallet::error]
pub enum Error<T> {
    ...
}
```

Reply:
To tell the truth no real reason. Currently this function is not called anywhere except tests so I would like to refactor it within the next task.